ast/compile.go

```diff
@@ -4591,14 +4588,49 @@ func validateTarget(c *Compiler, term *Term) *Error {
 			if child := node.Child(ref[len(ref)-1].Value); child != nil {
 				for _, value := range child.Values {
 					if len(value.(*Rule).Head.Args) > 0 {
-						return NewError(CompileErr, term.Loc(), "with keyword cannot replace functions")
+						// TODO(sr): UDF
+						return false, NewError(CompileErr, target.Loc(), "with keyword cannot replace rego functions")
 					}
 				}
 			}
 		}
+	case isInputRef(target): // ok, valid
+	case isBuiltinRefOrVar(c.builtins, target):
+		ref, ok := value.Value.(Ref)
+		if !ok {
+			return false, NewError(CompileErr, target.Loc(), "with keyword replacing built-in function: value must be a reference to a function")
+		}
+
+		node := c.RuleTree
+		for _, r := range ref {
+			child := node.Child(r.Value)
+			if child == nil {
+				return false, NewError(CompileErr, target.Loc(), "with keyword replacing built-in function: value must be a reference to a function")
+			}
+			node = child
+		}
+
+		bi := c.builtins[target.Value.String()] // safe because isBuiltinRef checked this
+
+		if bi.Relation {
+			return false, NewError(CompileErr, target.Loc(), "with keyword replacing built-in function: target must not be a relation")
+		}
 
+		for _, value := range node.Values {
+			arity := len(value.(*Rule).Head.Args)
+			biArity := len(bi.Decl.Args())
+			switch {
+			case arity == 0 && biArity != 0:
+				return false, NewError(CompileErr, target.Loc(), "with keyword replacing built-in function: referenced value must be a function")
+			case arity != biArity:
+				return false, NewError(CompileErr, target.Loc(), "with keyword replacing built-in function: referenced value function must have same arity (have %d, want %d)", arity, biArity)
+			}
+		}
+		return false, nil
+	default:
+		return false, NewError(TypeErr, target.Location, "with keyword target must reference existing %v, %v, or a built-in function", InputRootDocument, DefaultRootDocument)
```

Comment:
I was somewhat surprised to not be able to:

1. Mock a builtin function with another builtin function
2. Mock a builtin function with a value 

It's fine if we don't support these out of the gate but here are two use cases that came to mind:

1. Replace a built-in function w/ another that has less strict validation; if one function was a superset of the another function, that could be useful for test purposes (e.g., temporarily replace jwt.decode_strict with jwt.decode?)

2. Replace a built-in function like http.send w/ a static value; this is just a convenience thing, e.g., `p with http.send as {"body": {"foo": "bar"}}` is nice and convenient--no need to define a function unnecessarily. As a side note, I noticed this does work today on 0-arity functions, e.g., `opa.runtime() with opa.runtime as {"Foo": "Bar"}`

Reply:
Re 2: yeah I've come across this the other way around: `with opa.runtime as foo` works for both `foo = 3` and `foo() = 3`. 

But agreed, it makes sense to introduce this shortcut. :+1:

Re this and (2) -- the question is if it should become an immediate follow-up or if it's somewhat less important than mocking user-defined functions.

I'm leaning towards mocking user-defined functions _first_ (and getting _this_ merged quickly)... i.e. first unblock the use case, then optimize UX. WDYT?

Reply:
Thinking about this a little more, I think (2.) should better be part of this slice. I'd imagine this to be what neatly solves most testing needs right away. (Don't know why it hasn't occurred to me before, I was caught up with inline definitions and how to handle mocks that depend on the args there.)

I'll add static return values tomorrow 👍

Reply:
"tomorrow" (as of yesterday) ➡️  "tomorrow" (as of today)